File: problem-of-the-day/day05/Merge K Sorted List.py
class Solution(object):
    def mergeKLists(self, lists):
        # Base cases: if the input list is empty or contains only one list
        if not lists:
            return None
        if len(lists) == 1:
            return lists[0]

        # Split the list into two halves
        mid = len(lists) // 2
        l, r = self.mergeKLists(lists[:mid]), self.mergeKLists(lists[mid:])

        # Merge the two halves
        return self.merge(l, r)
    
    def merge(self, l, r):
        # Initialize a dummy node to help with merging
        dummy = p = ListNode()
        
        # Merge the two lists
        while l and r:
            if l.val < r.val:
                p.next = l
                l = l.next
            else:
                p.next = r
                r = r.next
            p = p.next
        
        # Attach the remaining elements, if any
        p.next = l or r
        
        # Return the merged list, which starts at dummy.next
        return dummy.next

File: problem-of-the-day/day05/Merge K Sorted List.cpp

class Solution {
public:
    ListNode* mergetwo(ListNode* h1, ListNode* h2)
    {
        ListNode* dummy = new ListNode(-1);
        ListNode* ptr = dummy;
        while(h1 && h2) 
        {
            if(h1->val < h2->val)
            {
                ptr->next = h1;
                h1 = h1->next;
            }
            else
            {
                ptr->next = h2;
                h2 = h2->next;
            }
            ptr = ptr->next;
        }

        // to check if any list is remaining, then adding the rest of the remaining list to dummy list whih is our answer
        if(h1!=NULL) ptr->next = h1; 
        if(h2!=NULL) ptr->next = h2;
        return dummy->next;
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()) return NULL;
        ListNode* temp = lists[0];
        if(lists.size()>1)
        {
            for(int i=1;i<lists.size();i++)
            {
                temp = mergetwo(temp,lists[i]);
            }
        }
        return temp;
    }
};